### PR TITLE
Pack native libs in .net core fashion

### PR DESCRIPTION
**Pull request by huoyaoyuan**
**Warning: This PR won't pass CI before merge**, unless adding a temporary workaround commit. You can do all the checks at local to **validate it does work**.

I've done the validations on Windows. Since there's no runtime gap unlike Mono comparing with .Net Framework, I except mac and Linux to simply success.

## Motivation
.NET Core introduces a better way to [configure native dependencies via package](https://docs.microsoft.com/en-us/nuget/create-packages/supporting-multiple-target-frameworks#architecture-specific-folders). Now legacy technologies get dropped and .NET Core is the only supported runtime for desktop, it's time to reform from various workarounds and get served by toolset conventions. The Xamarin builds ships totally different assets and won't be affected by this.

## Changes
The native libraries located in `runtimes/<rid>/native/` are well-handled by the dotnet sdk. No more need to manually setting path/naming conventions for them.

Calls to `SetDllDirectory` for Windows will be removed from framework.
Linux users no longer require to set the environment variable `LD_LIBRARY_PATH`.

`dotnet publish` will handle them properly. Deploy script should be updated to depend on it; old implications about file layout may break.
It will remove other OS's assets when packaging for specific OS. I can see `.dll` and `.so` in macOS package now.

Framework package changes to targeting both .NET Core and .NET Standard, and reference native libs package only for .NET Core target.
This can prevent miss-installing for the native package, as it's only designed for .NET Core.
Cross-framework targeting will be required further for framework, because .NET Core will be the only framework that's null-annotated in 1 year.

Size of framework NuGet package will be duped, because it contains assemblies for both framework. It's no harmful as the package is small enough today (~2.5M), and will only affect development machine.

### Why requiring manual validation
Though most parts of `dotnet` toolset handle it well, **`ProjectReference` doesn't know about native dependencies.** It will copy the files to the correct layout, but missing `project.assets.json` and `.deps.json` entries for them. This will result in `DllNotFoundException` at runtime, because CLR won't just load anything inside `runtimes\<rid>\native` folder. It only load known dependencies.

I have spent a whole day investigating if I can inject something into `.deps.json`. The task creating it accepts some dependency lists, but they won't be used. It follows exactly `project.assets.json`, which is generated by one concrete task. Copying them directly aside executable will have other problems too.

Since native libs seldom get updated, and ships in different pace comparing to framework, I think the best solution is to **only consume it via package reference**.
If the package following new convention hasn't be published to NuGet gallery, it can't be consumed by CI since we don't have nightly feeds like .NET official. Thankfully, NuGet provides capability to test package consuming at local machine.

### Steps to validate
Based on the latest commit.
1. Create a package with higher version than stable.
```
cd osu.Framework.NativeLibs
dotnet pack -p:version=9999.9.9
```

2. Put the package into local feed
Assuming directory `~\.nuget\hook\`
```
dotnet nuget push .\bin\Debug\ppy.osu.Framework.NativeLibs.9999.9.9.nupkg -s ~\.nuget\hook\
```
or
```
nuget.exe push .\bin\Debug\ppy.osu.Framework.NativeLibs.9999.9.9.nupkg -Source ~\.nuget\hook\
```

3. Update the package reference in framework project to using the higher version, and restore it with local feed
You can it directly in Visual Studio's NuGet setting, or restore using
```
 dotnet restore -s ~\.nuget\hook\
```
It's recommended to close your IDE and clean all the `bin` and `obj` folder before restore to ensure they're clean.

4. Build SampleGame.Desktop and desktop tests.
Check `SampleGame.Desktop.deps.json`. Search for `bass.dll`, `libbass.so` and `libbass.dylib`, you should find entries for them.
If there are native assets out of `runtime` folder, that's directly put outside or in `x86`/`x64`/`x86-64` folder, remove them to avoid misleading.

5. Run the sample game and visual tests. You shouldn't suffer `DllNotFoundException`.

6. Switch to the game repo. Install the latest package using
```
cd osu.Desktop
dotnet add package -v 9999.9.9 -s ~\.nuget\hook\ ppy.osu.Framework.NativeLibs
```
Verify desktop game also works. The new package is compatible with old framework, because workarounds are simply no-op when CLR can locate them.

7. Verify `dotnet publish -r` also works.
Be aware that if runtime identifier are specified, native assets will be put out directly, not in `runtimes` folder.

8. Don't forget to clean your NuGet global cache at `~\.nuget\packages\ppy.osu.Framework.NativeLibs\9999.9.9`.

### CODE REVIEW

File: osu.Framework.NativeLibs/osu.Framework.NativeLibs.csproj [hunk 1]
```diff
@@ -12,16 +12,20 @@
...
+    <NativeLibs Include="$(MSBuildThisFileDirectory)runtimes\**\*.so" />
+    <NativeLibs Include="$(MSBuildThisFileDirectory)runtimes\**\*.a" />
+    <NativeLibs Include="$(MSBuildThisFileDirectory)runtimes\**\*.dylib" />
     <None Include="@(NativeLibs)">
       <Pack>true</Pack>
-      <PackageCopyToOutput>true</PackageCopyToOutput>
-      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+      <PackagePath>runtimes</PackagePath>
+    </None>
+    <None Include="_._">
```
Comment (miterosan) [inline osu.Framework.NativeLibs/osu.Framework.NativeLibs.csproj]: What is this for?

Comment (huoyaoyuan) [reply osu.Framework.NativeLibs/osu.Framework.NativeLibs.csproj]: See https://docs.microsoft.com/en-us/nuget/reference/errors-and-warnings/nu5128#scenario-2

**Comment by smoogipoo** [comment]
Interesting! How come this doesn't build both netstandard and netcoreapp when building the desktop project? Was that a recent change? iirc I originally multi-targetted with the move to netstandard, but it used to compile all targets back then.

File: osu.Framework.NativeLibs/osu.Framework.NativeLibs.csproj (hunk 1 above)
Comment (smoogipoo) [reply osu.Framework.NativeLibs/osu.Framework.NativeLibs.csproj]: Why can't the `_._` file be directly placed inside `lib/netcoreapp3.0`? Seems like it should be the exact same thing but gives me the same warnings...

**Comment by huoyaoyuan** [comment]
> How come this doesn't build both netstandard and netcoreapp when building the desktop project? 

Seems that transitive `Build` task will only build the dependent one; `dotnet build` or `Pack` will build them both.

> Was that a recent change?

I introduced in this PR, because this **package layout** is only designed for netcoreapp.
If targeting other framework, the user will need to handle native assets manually. They can still use `<PackageTargetFallback>`, but should not get the package automatically.

> I originally multi-targetted with the move to netstandard, but it used to compile all targets back then.

Multi-targeting will be required soon when we enable nullable reference types. Looking forward, I think it's OK to do this earlier.

File: osu.Framework.NativeLibs/osu.Framework.NativeLibs.csproj (hunk 1 above)
Comment (huoyaoyuan) [reply osu.Framework.NativeLibs/osu.Framework.NativeLibs.csproj]: > directly placed

Do you mean not using `<PackagePath>`? Default packaged content will be added `contentFiles/` at the start of path. I always use https://github.com/NuGetPackageExplorer/NuGetPackageExplorer to check the actual package layout.

**Comment by smoogipoo** [comment]
So... Why won't this pass CI before merge? The issues that appveyor lists are actual issues as far as I'm seeing - code was altered and usings aren't required. Not sure what the `TypeExtensions` error is all about though.

**Comment by huoyaoyuan** [comment]
Tests cannot run because workarounds of old package layout are removed. I sayed that when I was trying to make `ProjectReference` work.
After the investigation I did, the workarounds can still exist because they are no harmful to the new package layout. Once new layout are published, they can be safely removed then.

**Comment by huoyaoyuan** [comment]
The inspection on `TypeExtensions` should be a bug of 2019.2.1. Once I updated to 2019.2.3 it disappears.

**Comment by huoyaoyuan** [comment]
Force pushed to remove the commit that removes old workaround.
Now it should pass CI, but **CI cannot inform what will happen**. Because no `ProjectReference`, every later update to native libraries should be tested manually, using the instructions I described.

**Comment by smoogipoo** [comment]
Personally I think this is fine, both the duping of the osu.Framework assembly and the manual validation of native libs.
Thoughts @peppy ?

File: osu.Framework.NativeLibs/osu.Framework.NativeLibs.csproj (hunk 1 above)
Comment (peppy) [reply osu.Framework.NativeLibs/osu.Framework.NativeLibs.csproj]: Can a comment be added in-line explaining this? Just the msdoc link may be enough.

Comment (huoyaoyuan) [reply osu.Framework.NativeLibs/osu.Framework.NativeLibs.csproj]: Sure. I often find things in `.nuget` to decompile and got already very familiar to `_._`, but others may not.

**Pull request merged**